lib/internal/errors.js

```diff
@@ -64,6 +65,16 @@ const prepareStackTrace = (globalThis, error, trace) => {
     return f(error, trace);
   }
 
+  for (let l = trace.length - 1; l >= 0; l--) {
+    const fn = trace[l].getFunctionName();
+    if (fn != null && fn.startsWith('__node_internal_hidden_')) {
```

Comment:
This is currently run on all errors, not only on Node.js errors of which some stack frames should be hidden. As such, it's a general overhead that we should not introduce. Instead, let's verify that this code path is only triggered for Node.js errors.

Reply:
This is true, I am at loss on how to do it without impacting the main path.

Reply:
Maybe execute it for all errors triggered everywhere in Node.js - but not for user code - by detecting if it originated in Node.js - I think this is doable without changing ```hideStackFrames```

Reply:
> This is currently run on all errors, not only on Node.js errors of which some stack frames should be hidden. As such, it's a general overhead that we should not introduce. Instead, let's verify that this code path is only triggered for Node.js errors.

It's a very good point. We should only enable the code path for built-in errors.

Reply:
Done, the check is for captureLargerStackTrace as innermost stack element.
when processing an async exception, prepareStackTrace is called without the ```Error``` object, so there are not many other solutions

Reply:
We have nested calls in some of validation functions, where only the outer function is wrapped with `hideStackFrames()`. For example, check `validateParameters` function from `lib/internal/crypto/hkdf.js`. If one of those nested validation functions (say, `validateString`) throws, the check here:
https://github.com/nodejs/node/pull/35644/commits/c205a80e9194153f3bb4203a21ef5b3f61c8df49#diff-670bf55805b781d9a3579f8bca9104c04d94af87cc33220149fd7d37b095ca1cR70
won't hold anymore. So, the user will get non-sanitized stack trace in the error.

I'd say, that the current approach to avoid the loop for user errors seems to be fragile and error-prone.

Reply:
If processing a Node error with internals to hide, that last function on the stack will always be `captureLargerStackFrame`

Reply:
> If processing a Node error with internals to hide, that last function on the stack will always be `captureLargerStackFrame`

Yes, you're right. It may be a good idea to leave a comment around the check and mention that the check will hold for built-in errors.